File: 15/activity15 Java/src/com/gdb/domain/AccountFactory.java
package com.gdb.domain;

import com.gdb.exceptions.InvalidAccountTypeException;
import com.gdb.exceptions.InvalidAgeException;
import com.gdb.exceptions.MinimumBalanceViolationException;

public class AccountFactory {
    public static IAccount createAccount(String accountType, int accountNumber, String name, int age, double initialBalance, int tenureYears)
            throws InvalidAgeException, InvalidAccountTypeException, MinimumBalanceViolationException {
        if (accountType == null) {
            throw new InvalidAccountTypeException("Account type cannot be null");
        }
        String t = accountType.toUpperCase().replace(" ", "").replace("_", "");
        IAccount acc;
        if ("SAVINGS".equals(t)) {
            acc = new SavingsAccount(accountNumber, name, age, initialBalance, tenureYears);
        } else if ("CURRENT".equals(t)) {
            acc = new CurrentAccount(accountNumber, name, age, initialBalance, tenureYears);
        } else if ("FIXEDDEPOSIT".equals(t) || "FD".equals(t)) {
            acc = new FixedDepositAccount(accountNumber, name, age, initialBalance, tenureYears);
        } else if ("SALARY".equals(t)) {
            acc = new SalaryAccount(accountNumber, name, age, initialBalance, tenureYears);
        } else {
            throw new InvalidAccountTypeException("Unsupported account type: " + accountType);
        }

        if (initialBalance < acc.getMinimumBalance()) {
            throw new MinimumBalanceViolationException(
                String.format("Initial balance Rs. %,.2f is below minimum required Rs. %,.2f",
                    initialBalance, acc.getMinimumBalance())
            );
        }
        return acc;
    }

    public static IAccount createAccount(String accountType, int accountNumber, String name, int age, double initialBalance)
            throws InvalidAgeException, InvalidAccountTypeException, MinimumBalanceViolationException {
        return createAccount(accountType, accountNumber, name, age, initialBalance, 0);
    }
}

File: 15/activity15 Python/gdb/domain/account_factory.py
# gdb/domain/account_factory.py
from gdb.domain.iaccount import IAccount
from gdb.domain.savings_account import SavingsAccount
from gdb.domain.current_account import CurrentAccount
from gdb.domain.fixed_deposit_account import FixedDepositAccount
from gdb.domain.salary_account import SalaryAccount
from gdb.exceptions.invalid_account_type_exception import InvalidAccountTypeException
from gdb.exceptions.minimum_balance_violation_exception import MinimumBalanceViolationException

class AccountFactory:
    """Factory creating appropriate IAccount instances."""

    @staticmethod
    def create_account(account_type: str, account_number: int, name: str, age: int, initial_balance: float, tenure_years: int = 0) -> IAccount:
        t = account_type.upper().replace(" ", "").replace("_", "")
        acc: IAccount
        if t == "SAVINGS":
            acc = SavingsAccount(account_number, name, age, initial_balance, tenure_years)
        elif t == "CURRENT":
            acc = CurrentAccount(account_number, name, age, initial_balance, tenure_years)
        elif t in ("FIXEDDEPOSIT", "FD"):
            acc = FixedDepositAccount(account_number, name, age, initial_balance, tenure_years)
        elif t == "SALARY":
            acc = SalaryAccount(account_number, name, age, initial_balance, tenure_years)
        else:
            raise InvalidAccountTypeException(f"Unsupported account type: {account_type}")

        if initial_balance < acc.get_minimum_balance():
            raise MinimumBalanceViolationException(
                f"Initial balance Rs. {initial_balance:,.2f} is below minimum required Rs. {acc.get_minimum_balance():,.2f}"
            )
        return acc

    @staticmethod
    def createAccount(account_type: str, account_number: int, name: str, age: int, initial_balance: float, tenure_years: int = 0) -> IAccount:
        return AccountFactory.create_account(account_type, account_number, name, age, initial_balance, tenure_years)
